numpy/core/tests/test_deprecations.py

```diff
@@ -1100,3 +1100,41 @@ def check_parametrize(base, power, expected):
                 count += 1
             assert_(count == 3)
         self.assert_deprecated(_test_parametrize)
+
+
+class TestStringPromotion(_DeprecationTestCase):
+    # Deprecated 2020-12-19, NumPy 1.21
+    warning_cls = FutureWarning
+    message = "Promotion of numbers and bools to strings is deprecated."
+
+    @pytest.mark.parametrize("dtype", "?bhilqpBHILQPefdgFDG")
+    @pytest.mark.parametrize("string_dt", ["S", "U"])
+    def test_deprecated(self, dtype, string_dt):
+        self.assert_deprecated(lambda: np.promote_types(dtype, string_dt))
+
+        # concatenate has to be able to promote to find the result dtype:
+        arr1 = np.ones(3, dtype=dtype)
+        arr2 = np.ones(3, dtype=string_dt)
+        self.assert_deprecated(lambda: np.concatenate((arr1, arr2), axis=0))
+        self.assert_deprecated(lambda: np.concatenate((arr1, arr2), axis=None))
+
+        # coercing to an array is similar, but will fall-back to `object`
+        # (when raising the FutureWarning, this already happens)
+        self.assert_deprecated(lambda: np.array([arr1[0], arr2[0]]),
+                               exceptions=())
```

Comment:
Something I randomly realized: I could change this so that it it actually raises the `FutureWarning` here. That is: We could differentiate between "no promotion found/possible" and an *error* while trying to promote).
But, I am not sure it makes things better here, and another solution for that distinction is to create a `PromotionError` type (assuming that invalid promotions are unlikely to be ever speed relevant).

Reply:
It seems we would want to make sure a warning is noticed when promotion fails. Should we merge as-is and enhance in another PR or add it here?

Reply:
Sorry, just to be clear (just to be sure), this affects:
```
warnings.simplefilter("error", FutureWarning)
np.array([0, "a"])
```
which would "opt in" to the new behaviour currently.  Happy to change it. But maybe the distinction for custom errors doesn't matter too much.  If we introduce a `PromotionTypeError` to allow custom errors (if that even makes sense for an operator), we could still catch `PromotionTypeError` explicitly.

Reply:
I forgot that this also affects, the second step, which would be the one that finds the correct unit for two different datetimes.  If you think its worthwhile for this warning, I can look into it. Otherwise, I think swapping out to an explicit error type might be more elegant at some point (if we ever care). Since we should also need to say that `__common_instance__ ` should return NotImplemented for such impossible cases, which seems not really that great to me.